### DISCUSSION TITLE
Is this considered to be an okay practice in nextjs middleware?

### DISCUSSION THREAD

**Discussion by josephp27** [vercel/next.js#60804] [Help]
### Summary

Hi, is this considered to be an ok practice? I am only doing this on my api routes and would like to reduce code overhead by authenticating and passing the user to the downstream api. Obviously, I don't want a security issue to happen. However, this afaik this is inject internally and would not be susceptible to a mitm attack

Thanks!
```js
import { createMiddlewareSupabaseClient } from "@supabase/auth-helpers-nextjs";
import { NextRequest, NextResponse } from "next/server";

export async function middleware(req: NextRequest) {
  const res = NextResponse.next();

  const supabase = createMiddlewareSupabaseClient({ req, res });

  const {
    data: { session },
  } = await supabase.auth.getSession();

  if (!session) {
    return new Response(null, { status: 401, statusText: "Not Authorized" });
  }

  // add the user to the request
  const requestHeaders = new Headers(req.headers);
  requestHeaders.set("X-User", JSON.stringify(session.user));

  const response = NextResponse.next({
    request: {
      headers: requestHeaders,
    },
  });

  return response;
}

// match anything that is an api
export const config = {
  matcher: "/api/:limits*",
};
```

**Comment by icyJoseph** [comment]
This should be fine, as far as the architecture goes, because matched routes, will execute the middleware first. In other words, as you have it, I don't know of a way to hit the underlying API Route handler, without hitting the middleware. 

It wouldn't hurt to do some basic checks in your API handlers though.

**Reply by josephp27** [reply]
Hey, thanks for your reply! What about between the middleware and the api. Is there a way for a bad actor to rewrite the user header im writing in the middleware step?

**Reply by icyJoseph** [reply]
Mmm I think the header might be visible to the end user.

**Reply by josephp27** [reply]
that's no problem if they can read it, since it is their information. I just don't want them rewriting the user id and accessing another user's resource. hopefully that makes sense

**Reply by icyJoseph** [reply]
Right, but you always derive that from the supabase call, and since middleware runs in all api requests, it should be fine. Though, if a page uses this information, then it wouldn't be covered by the middleware guard.

It is always a bit iffy to give a seal of approval when it comes to security without looking at the whole picture, and even then, there are many other things that could go wrong. In this case though, this feels like one of the use cases middleware was created for.
